### ISSUE TITLE
bug of performance drop

### ISSUE THREAD

**Issue by yongkangchen** [LuaJIT/LuaJIT#122]
There are two tests, the only difference is the execute order, but get the different performance.
### Code of normal result:
```
require "vector3"
collectgarbage()
test3() -- output: test3/lua 0.000874
collectgarbage()
test2() -- output: test2/lua 0.063899
```

###  Code of bad result:
```
require "vector3"
collectgarbage()
test2() -- output: test2/lua 0.058866
collectgarbage()
test3() -- output: test3/lua 0.253925
```

#####  Code of vector3:
```
local Vector3 = { __typename = 'Vector3' }
do
        local setmetatable = setmetatable
        local I = { __typename='Vector3' }

        Vector3.__call = function(t, x, y, z)
                return setmetatable({x or 0, y or 0, z or 0}, I)
        end
        setmetatable(Vector3, Vector3)
end

function test2()
        local start = os.clock()
        for i = 1,200000 do
                local up = Vector3(0,1,0)
                tostring(up)
        end
        print("test2/lua " .. (os.clock() - start));
end

function test3()
        local start = os.clock()
        for i = 1, 2000000 do
                Vector3(i, i, i)
        end
        print("test3/lua " .. (os.clock() - start));
end
```

**Comment by yongkangchen** [comment]
ps: This bug both exist in mac, linux, win. And both luajit-2.0.4 and luajit-2.1.0beta1

**Comment by mraleph** [comment]
I can't reproduce this behavior (neither on Mac nor on Linux). Here is what I get on Linux:

https://gist.github.com/mraleph/7a75daee13ad783d8f78

Could you include `-jdump` output for fast and slow runs?

**Comment by fsfod** [comment]
test3 turns into loop that does nothing because the vector is not stored to something outside the loop and test2 will constantly enter and exit the JIT every loop iteration because its stitched at the tostring call.

**Comment by yongkangchen** [comment]
@mraleph `-jdump` output : https://gist.github.com/yongkangchen/51be7b9b70fc73dc43d4

**Comment by yongkangchen** [comment]
However, if I change code of `Vector3.__call` to:
```
        Vector3.__call = function(t, x, y, z)
                local v = {x or 0, y or 0, z or 0}
                setmetatable(v, I)
                return v
        end
```
this performance drop disappeared.

**Comment by yongkangchen** [comment]
Also this could avoid this performance drop too:
```
        Vector3.__call = function(t, x, y, z)
                local v = setmetatable({x or 0, y or 0, z or 0}, I)
                return v
        end
```
so I guess jit skip on function with tail call?

**Comment by mraleph** [comment]
Thanks for the dump. I can explain what happens (and I can actually reproduce on v2.1.0-beta1, I think yesterday I was running the wrong build - sorry about that).

If you call `test2` first then JIT first tries to compile a trace starting at the loop (which it can't because `NYI: unsupported variant of FastFunc tostring`) then it repeatedly tries to compile a trace starting at the function entry of  `Vector3.__call` - which it can't because of the tail call. Function entry then gets **blacklisted**. Then you call `test3` and when it tries to compile a trace starting at the loop it hits blacklisted function entry and aborts.

The problem is not reproducible on the HEAD of v2.1 branch because trace stitching enables compilation of the loop trace starting inside `test2` and as the result `Vector3.__call` is not blacklisted.

**Comment by zhaozg** [comment]
intresting

**Comment by yongkangchen** [comment]
@mraleph so, I guess if there is any NYI function call in a large count loop, tail call in loop will be blacklisted forever?

**Comment by yongkangchen** [comment]
If so, can you explain why? :)

**Comment by MikePall** [comment]
Closing since it's fixed in 2.1
